Blog timelock refreshing via Payjoin (#21)
payjoin/payjoindevkit.org: docs/_blog/liana-payjoin.md

---
title: "Refreshing timelocked UTXOs with Payjoin"

> description: "Using Payjoin to Preserve Privacy and Lower Fees in Timelocked Wallets"

Reviewer: Title and description should definitely mention Liana and perhaps inheritance and other keywords the target audience might search to find this.

Yoast has a decent article on writing one: https://yoast.com/meta-descriptions/

date: "2025-04-09"
authors:
  - Armin Sabouri
tags:
  - privacy
  - integrations
---

> The highest form of self-custody involves using multiple geographically distributed hardware keys. But what happens if you lose a quorum of those keys? Many users turn to paid services like [Casa](https://casa.io/) or [Unchained](https://unchained.com/) to have a third party manage a subset of their keys in a self-custodial way. [Liana](https://wizardsardine.com/liana/) addresses this recovery problem differently. It allows users to opt into a spending policy that includes a time-locked recovery path.

Reviewer: This is a very in-the-know bitcoiner orientation. I'd prefer if it were oriented around the benefit for inheritence specifically and did not mention Casa and Unchained.

<img alt="Liana Timelocked UTXOs" src="../assets/liana-coins.png" />

Reconciliation, however, introduces a privacy risk. It creates an on-chain footprint that allows surveillance firms to estimate a user's wallet balance and transaction history. This is primarily due to the [common input ownership heuristic](https://en.bitcoin.it/wiki/Common-input-ownership_heuristic), which assumes all inputs in a transaction belong to the same entity. Combined with unique transaction patterns that act like fingerprints, these heuristics allow on-chain analysts to cluster a user's UTXOs and glean sensitive information.

## Payjoin as a Privacy-Preserving Refresh

> 
> This is where Payjoin becomes valuable. It allows Liana users to refresh or reconcile soon-to-expire UTXOs during normal payments. The same method can apply to self-spends. If a user isn’t expecting incoming payments, they can still use Payjoin the next time they send a transaction. Whether as sender or receiver, users benefit from reduced fees and a more obfuscated transaction graph—preserving privacy while maintaining the integrity of their recovery policy.

Reviewer: ```suggestion

This is where Payjoin becomes valuable. It allows Liana users to refresh or reconcile soon-to-expire UTXOs during normal payments. The same method can apply to self-spends. If a user isn’t expecting incoming payments, they can still use Payjoin the next time they send a transaction. Whether as sender or receiver, users benefit from reduced fees and a more private transaction graph while maintaining the integrity of their recovery policy.
```

Reviewer: obfuscate literally means to make dark. I don't think that's the vibe

<img alt="Liana Payjoin" src="../assets/liana-batch-payment.png" />

## Caveats

It's important to note that there is no privacy gain when using Payjoin with P2WSH descriptors unless both the sender and receiver share the same or structurally similar descriptor. This is because the spending script is revealed on-chain, allowing observers to fingerprint the wallet based on its script template. If the participants use different descriptors, their inputs become trivially distinguishable, defeating the purpose of Payjoin. Taproot-based descriptors using Miniscript do not suffer from this issue. With Taproot, key-path spending hides the recovery script entirely unless deliberately revealed, enabling privacy-preserving Payjoin transactions even between users with different wallet policies.

> 

Reviewer: "privacy gain" is another term I like to avoid. I prefer something like "...the ability to break common input ownership is limited when..."

---

> title: "Automatically Refresh Inheritance Time-Locks with Payjoin"

Reviewer: I've seen "Timelock" [everywhere](https://en.bitcoin.it/wiki/Timelock) and never Time-lock. Is that standard / search optimized?

Reviewer: either that or [with space](https://bitcoinops.org/en/topics/htlc/)

Reviewer: more [timelock](https://blog.casa.io/bitcoin-multisig-time-locking-challenges/)

Reviewer: found a [Time-locked](https://thebitcoinmanual.com/btc-finance/tlc/) instance, but I'll let you decide how this matches in authority compared to the prior comments

Author: I'd prefer "timelock" no space or hyphen.

Author: And similarly "time-locked" -> "time locked" as two words

> title: "How Liana Integrates Payjoin to Reduce Fees and Enhance Privacy"

Author: This comment is outdated even tho it was made 3 minutes ago :) 
The title is : "Automatically Refresh Inheritance Time-Locks with Payjoin"

description: "Liana combines Payjoin and time locked inheritance to help you keep your Bitcoin private now and safe for those who’ll need it later."
date: "2025-04-10"
authors:
  - Armin Sabouri
tags:
  - privacy

>   - integrations

Reviewer: cut-through is another appropriate tag

Author: Done

---

[Liana](https://wizardsardine.com/liana/) is a Bitcoin desktop wallet that enables users to set up time-locked inheritance policies, ensuring their funds can be passed on to loved ones. To prevent unintentional recovery path activation, these time-locked coins require periodic refreshing. Liana’s Payjoin integration enables users to refresh timelocks as they receive payments, boosting privacy and cutting costs.

## Privacy Leak Explained

> Inheritance planning is a crucial yet often overlooked part of Bitcoin self-custody. While Bitcoin offers unmatched financial sovereignty, it introduces [unique challenges](https://wizardsardine.com/blog/the-state-of-bitcoin-inheritance-planning/) when it comes to passing on funds to future generations. Liana addresses this with inheritance policies that activate only after a predefined delay called a "time-lock". For example, a user might control a primary key for daily use, while a recovery key held by a beneficiary is time-locked for future access.

Reviewer: if you're taking the suggestion to introduce terminology ("time-lock") before using it, that needs to be done before it's done in the first sentence of the article, especially if it's used in the title

Author: Good point. Moved the timelock introduction into the introduction

> However, there's a catch: most users don't frequently spend their Bitcoin. [A 2024 study](https://www.sciencedirect.com/science/article/pii/S0378437124008045) shows the age of unspent coins follows a power-law distribution. In other words, coins are either spent quickly or left untouched for long periods. [Data also suggests](https://charts.bitbo.io/hodl-waves/) that most coins remain unspent for months or years. To avoid unintentionally triggering a recovery path, users often resort to self-spending or reconciliation to reset timelocks.

Reviewer: show => cite the figure?

Reviewer: is self-spend different from reconciliation? Why 2 pieces of vocab?

Author: Fixed. And cited the more relevant figure. Thanks for pointing that out

<img alt="Liana time locked coins" src="../assets/liana-coins.png" />

But this reconciliation comes at a cost to privacy. These transactions create an identifiable on-chain footprint, making it easier for on-chain surveillance to estimate a user's balance and transaction history. This is primarily due to the [common input ownership heuristic](https://en.bitcoin.it/wiki/Common-input-ownership_heuristic), which assumes all inputs in a transaction belong to the same entity. Combined with unique transaction patterns that act like fingerprints, these heuristics allow on-chain analysts to cluster a user's addresses and glean sensitive information.

## Payjoin as a Privacy-Preserving Refresh

> This is where Payjoin shines. It allows Liana users to refresh near-expiry coins during normal payments. For example, when a user receives a payment via Payjoin, they can opportunistically include expiring coins to refresh their timelocks. This same logic can apply to self-spends—users can initiate a Payjoin when sending a transaction, embedding refresh into routine activity.

Reviewer 2: The emdash doesn't appear (in markdown at least), so a space would help
```suggestion
This is where Payjoin shines. It allows Liana users to refresh near-expiry coins during normal payments. For example, when a user receives a payment via Payjoin, they can opportunistically include expiring coins to refresh their timelocks. This same logic can apply to self-spends — users can initiate a Payjoin when sending a transaction, embedding refresh into routine activity.
```

Author: Good point!

Reviewer: you can use `&emdash;` in html instead

Author: Done

Whether as sender or receiver, users benefit from lower fees, improved privacy, and preservation of their inheritance policies.

<img alt="Liana Payjoin" src="../assets/liana-batch-payment.svg" />

## Caveats

> It’s worth noting that the common-input heuristic is not defeated when using Payjoin with P2WSH inputs—unless both participants share identical or structurally similar descriptors. This is because the spending script is revealed on-chain, allowing fingerprinting based on script templates.

Reviewer 2: Does the reader at this point know that Liana uses P2WSH inputs or what that is? Might be useful to at least mention that Liana uses those types of inputs and that P2WSH are used for multisig, with a link on P2WSH explaining what that is.

Author: Provided a couple links and a note on default inputs Liana uses

> However, Taproot inputs do not suffer from this issue. With Taproot’s key-path spending, the recovery script remains hidden unless explicitly revealed. This makes it possible to conduct Payjoin transactions between users with different wallet policies while breaking the common-input heuristic.

Reviewer 2: If Taproot inputs are the main type of inputs being used today (or are at least on track to be), this will strengthen the use case for payjoin in Liana

Author: They need to know b/c everything I've said about CIH is not mostly not true if you are not using taproot. Think of this section as more of a disclaimer. With that said, would you re-word or organize it differently?

Author: Also _I think_ taproot inputs are not default bc of hardware wallet support for tapminiscript descriptors. I know this from personal experience.  Can't find a chart to back this up unfortunately.

Reviewer 2: > Why does the reader at this point know or care about Taproot inputs?

Sorry I understand the reason for Taproot, I meant that the reader doesn't know the reason for Taproot if they aren't familiar with what Taproot is. I basically meant that we should briefly introduce Taproot (with a link preferably) as a newer input type that has properties which do not suffer from this issue. So perhaps something like:

```suggestion
It’s worth noting that the common-input heuristic is not defeated when using Payjoin with P2WSH inputs—unless both participants share identical or structurally similar descriptors. This is because the spending script is revealed on-chain, allowing fingerprinting based on script templates.

However, [Taproot inputs](https://thebitcoinmanual.com/articles/pay-to-taproot-p2tr), a newer input type, do not suffer from this issue. With Taproot’s key-path spending, the recovery script remains hidden unless explicitly revealed. This makes it possible to conduct Payjoin transactions between users with different wallet policies while breaking the common-input heuristic.
```

> However, Taproot inputs do not suffer from this issue. With Taproot’s key-path spending, the recovery script remains hidden unless explicitly revealed. This makes it possible to conduct Payjoin transactions between users with different wallet policies while breaking the common-input heuristic.

Reviewer 2: > With Taproot’s key-path spending

Generally prefer links to any technical jargon so the reader can learn more if they'd like. I'm usually opposed to assuming the reader knows what we mean beyond the basics. If it's not overly pedantic, I either like to spell it out or provide links to learn more.

Author: Very good point!

## Proof of Concept and Next Steps

During the MIT Bitcoin Hackathon, the Payjoin team developed a working proof of concept for both the Payjoin receiver and sender integrated with Liana. You can view the prototype [here](https://github.com/0xBEEFCAF3/liana/tree/payjoin-hackathon). In just 30 hours, the team successfully implemented the core functionality.

Looking ahead, we’re coordinating closely with the Liana engineering team to bring this integration to production. This involves rigorous testing, refining the transaction construction logic, and ensuring compatibility with Liana’s policy enforcement mechanisms. Together, we’ll work toward a robust, production-ready release that empowers users with a privacy-preserving, fee-efficient way to manage their recovery paths.

> This integration brings privacy and practicality to Bitcoin inheritance, showing how simple batching protocols like Payjoin can improve user experience without compromising privacy.

Reviewer 2: A lot of focus on the privacy aspects here, which is fine in this article and probably the major use-case for a Liana-type integration, but just noting that our emphasis if we care about adoption should be fee-and-efficiency savings.

Reviewer 2: What you have is good I think, I guess I figured this struck me as privacy-centric while reading it, though our goal in _general_ with payjoin is to move the marketing from "privacy is primary goal with time/fee savings as byproduct" to vice-versa. But in the Liana case I think what you have is good 👍

---
title: "Automatically Refresh Liana Inheritance Timelocks with Payjoin"
description: "Liana combines Payjoin and time locked inheritance to help you keep your Bitcoin private now and safe for those who’ll need it later."
date: "2025-04-10"
authors:
  - Armin Sabouri
tags:
  - privacy
  - integrations
  - cut-through
---

> [Liana](https://wizardsardine.com/liana/) is a Bitcoin desktop wallet that enables users to set up inheritance policies, ensuring their funds can be passed on to loved ones only after a predefined delay called a "timelock". To prevent unintentional activation of the inheritance policy, these time locked coins require periodic refreshing. Liana's Payjoin integration enables users to refresh timelocks as they receive payments, boosting privacy and cutting costs.
> 
> ## Privacy Leak Explained
> 
> Inheritance planning is a crucial yet often overlooked part of Bitcoin self-custody. While Bitcoin offers unmatched financial sovereignty, it introduces [unique challenges](https://wizardsardine.com/blog/the-state-of-bitcoin-inheritance-planning/) when it comes to passing on funds to future generations. Liana addresses this with inheritance policies that activate only after a timelock. For example, a user might control a primary key for daily use, while a recovery key held by a beneficiary is time locked for future access.
> 
> However, there's a catch: most users do not frequently spend their Bitcoin. A [2024 study](https://www.sciencedirect.com/science/article/pii/S0378437124008045) found that the age of unspent coins follows a power-law distribution (see Figure 3 that study). In practice, this means coins are typically either spent soon after receipt or held for long periods. [Data also suggests](https://charts.bitbo.io/hodl-waves/) that most coins remain unspent for months or years. To avoid unintentionally triggering a recovery path, users often resort to reconciliation of their coins to reset timelocks.
> 

Reviewer: ```suggestion
[Liana](https://wizardsardine.com/liana/) secures bitcoin such that after time passes, if and only if you haven't yet spent your coins, your heirs can spend them. This condition is called a "timelock" contract. To prevent unintentional activation of this inheritance policy, these timelocked coins require periodic refreshing by being spent into a new timelock. Liana's Payjoin integration automates timelock refresh as payments are received, boosting privacy and cutting costs.

Timelock contracts define some time in the future when funds can be spent by heirs. Before that, only the primary owner can spend the funds. However, there's a catch: most users do not frequently spend their Bitcoin. A [2024 study](https://www.sciencedirect.com/science/article/pii/S0378437124008045) found that the age of unspent coins follows a power-law distribution (see Figure 3 that study). In practice, this means coins are typically either spent soon after receipt or held for long periods. [Data also suggests](https://charts.bitbo.io/hodl-waves/) that most coins remain unspent for months or years. To avoid unintentionally triggering a recovery path, users often resort to reconciliation of their coins to reset timelocks.
```

I don't think the "Privacy leak explained" heading precedes an explanation of a privacy leak, so I removed it. If you still wanted to use the heading to break up the flow, the heading could follow the image.

Author: This grouping of paragraphs makes sense to me.

Author: Done

<img alt="Liana time locked coins" src="../assets/liana-coins.png" />

But this reconciliation comes at a cost to privacy. These transactions create an identifiable on-chain footprint, making it easier for on-chain surveillance to estimate a user's balance and transaction history. This is primarily due to the [common input ownership heuristic](https://en.bitcoin.it/wiki/Common-input-ownership_heuristic), which assumes all inputs in a transaction belong to the same entity. Combined with unique transaction patterns that act like fingerprints, these heuristics allow on-chain analysts to cluster a user's addresses and glean sensitive information.

## Payjoin as a Privacy-Preserving Refresh

> This is where Payjoin shines. It allows Liana users to refresh near-expiry coins during normal payments. For example, when a user receives a payment via Payjoin, they can opportunistically include expiring coins to refresh their timelocks. This same logic can apply to self-spends&mdash;users can initiate a Payjoin when sending a transaction, embedding refresh into routine activity.

Reviewer: ```suggestion
This is where Payjoin shines. When a user receives via Payjoin they get a chance to spend their own coins and tell the sender which outputs to create. In doing so Payjoin may spends timelocked coins to refresh their timelocks while adding funds to their stack. This same logic can apply to self-spends&mdash;users can initiate a Payjoin when sending a transaction, embedding refresh into routine activity.
```

I didn't make a change to the last sentence but I don't really get it. Does payjoin embed refresh into routine activity during spends that would not otherwise be possible? I don't think it does. In that scenario the privacy benefit is more relevant, isn't it?

Author: I thought about this more. I dont think we can make that claim and also not adhere to unnecessary input heuristic. More over if Liana lets you select unnecessary inputs during a spend then this point is moot.

> Whether as sender or receiver, users benefit from lower fees, improved privacy, and preservation of their inheritance policies.

Reviewer: "Whether _ as a sender or receiver" is missing a verb.

---

is preservation the correct word choice in this sentence? technically the policy is *maintained* by being renewed, because preserving it would allow it to expire. I prefer "maintenance" or "renewal"

Author: Ended up removing this sentence. Unnecessary transtion imo

<img alt="Liana Payjoin" src="../assets/liana-batch-payment.svg" />

## Caveats

> It's worth noting that the common-input heuristic is not defeated when using Payjoin with [P2WSH](https://river.com/learn/terms/p/p2wsh/) inputs&mdash;unless both participants share identical or structurally similar wallet policies. This is because the spending script is revealed on-chain, allowing fingerprinting based on script templates. Liana uses P2WSH by default to maintain compatibility with hardware wallets, which currently have limited support for [Taproot miniscript descriptors aka MiniTapscript](https://github.com/bitcoin/bitcoin/pull/27255).

Reviewer: ```suggestion
Note that the common-input heuristic is not defeated when using Payjoin with Liana's default [P2WSH](https://river.com/learn/terms/p/p2wsh/) inputs&mdash;unless both participants share similar wallet policies. Because the spending script is always revealed on-chain, it leaves a distinct fingerprint from other script types. Liana uses P2WSH by default to maintain compatibility with hardware wallets, which currently have limited support for [Taproot miniscript descriptors aka MiniTapscript](https://github.com/bitcoin/bitcoin/pull/27255).
```

similar includes identitcal

Author: Done

> However, [Taproot inputs](https://thebitcoinmanual.com/articles/pay-to-taproot-p2tr), a newer input type, do not suffer from this issue. With Taproot's key-path spending, the recovery script remains hidden unless explicitly revealed. This makes it possible to conduct Payjoin transactions between users with different wallet policies while breaking the common-input heuristic.

Reviewer: No however if you're concluding with the connection to taproot already (great transition otherwise)

```suggestion
[Taproot inputs](https://thebitcoinmanual.com/articles/pay-to-taproot-p2tr), which use a newer script type, do not suffer from this issue. With Taproot's key-path spending, the recovery script remains hidden unless it is invoked. Taproot makes it possible to conduct Payjoin transactions between users with different wallet policies and still break the common-input heuristic.
```

## Proof of Concept and Next Steps

During the MIT Bitcoin Hackathon, the Payjoin team developed a working proof of concept for both the Payjoin receiver and sender integrated with Liana. You can view the prototype [here](https://github.com/0xBEEFCAF3/liana/tree/payjoin-hackathon). In just 30 hours, the team successfully implemented the core functionality.

Looking ahead, we’re coordinating closely with the Liana engineering team to bring this integration to production. This involves rigorous testing, refining the transaction construction logic, and ensuring compatibility with Liana’s policy enforcement mechanisms. Together, we’ll work toward a robust, production-ready release that empowers users with a privacy-preserving, fee-efficient way to manage their recovery paths.

> This integration brings privacy and practicality to Bitcoin inheritance, showing how simple batching protocols like Payjoin can improve user experience without compromising privacy.

Reviewer: solid wrap

---
title: "Automatically Refresh Inheritance Time-Locks with Payjoin"
description: "Liana combines Payjoin and time-locked inheritance to help you keep your Bitcoin private now and safe for those who’ll need it later."
date: "2025-04-10"
authors:
  - Armin Sabouri
tags:
  - privacy
  - integrations
---

> <!-- create some spacing -->
> <h2></h2>
> 
> ## Introduction

Reviewer 3: Instead of this hack, we can remove the Introduction header which isn't necessary.

```suggestion
```

Author: Done!

[Liana](https://wizardsardine.com/liana/) is a Bitcoin desktop wallet that enables users to create time-locked inheritance policies, ensuring their funds can be passed on securely. However, managing these time-locked UTXOs requires periodic refreshing to avoid premature activation of recovery paths. By integrating Payjoin, Liana allows users to refresh UTXOs during routine payments—preserving privacy and saving on fees. Let’s explore how this combination works.

## Privacy Leak Explained

> Inheritance planning is a crucial yet often overlooked part of Bitcoin self-custody. While Bitcoin offers unmatched financial sovereignty, it introduces [unique challenges](https://wizardsardine.com/blog/the-state-of-bitcoin-inheritance-planning/) when it comes to passing on funds to future generations. Liana addresses this with time-locked inheritance policies that activate only after a predefined delay. For example, a user might control a primary key for daily use, while a recovery key held by a beneficiary is time-locked for future access.
> 

Reviewer: ```suggestion
Inheritance planning is a crucial yet often overlooked part of Bitcoin self-custody. While Bitcoin offers unmatched financial sovereignty, it introduces [unique challenges](https://wizardsardine.com/blog/the-state-of-bitcoin-inheritance-planning/) when it comes to passing on funds to future generations. Liana addresses this with inheritance policies that activate only after a predefined delay called a "timelock". For example, a user might control a primary key for daily use, while a recovery key held by a beneficiary is time-locked for future access.

```

Introduce jargon with plain speech and then feel free to use it like a wildman
